apps/src/blockly/addons/cdoConstantsProvider.js

```diff
@@ -0,0 +1,122 @@
+import GoogleBlockly from 'blockly/core';
+import {inputBlockTypes, outputBlockStyleTypes} from './cdoConstants';
+
+export default class CdoConstantsProvider extends GoogleBlockly.blockRendering
+  .ConstantProvider {
+  /**
+   * Get an object with connection shape and sizing information based on the
+   * type of the connection.
+   *
+   * @param connection The connection to find a shape object for
+   * @returns The shape object for the connection.
+   * @override
+   */
+  shapeFor(connection) {
+    const inputTypeShapeMap = {
+      [inputBlockTypes.SPRITE]: this.TRI_INPUT_OUTPUT,
+      [inputBlockTypes.BEHAVIOR]: this.ROUND_INPUT_OUTPUT,
+      [inputBlockTypes.LOCATION]: this.RECT_INPUT_OUTPUT
+    };
+    const outputStyleShapeMap = {
+      [outputBlockStyleTypes.SPRITE_TYPE]: this.TRI_INPUT_OUTPUT,
+      [outputBlockStyleTypes.BEHAVIOR_TYPE]: this.ROUND_INPUT_OUTPUT,
+      [outputBlockStyleTypes.LOCATION_TYPE]: this.RECT_INPUT_OUTPUT
+    };
+    const blockStyleName = connection.getSourceBlock().styleName_;
```

Comment:
I think the bug with the insertion markers is happening because insertion markers don't get the normal style name for the block they're a marker of. They are a different color so they don't have the same style. I did some testing and for insertion marker blocks, this comes up as `auto_#000000` or something similar instead of `sprite_blocks` etc. This causes this lookup to fail and you fall back to the puzzle shape.

You should probably use the same `connection.check_` value instead of trying to use the category of the block. For example, when debugging this I used text and number blocks since I don't have sprite and behavior blocks. But the `text_count` block outputs a number despite being in the text block category. If you went off the category of the blocks instead, you would have that block with a misleading shape. By making sure you are always basing it off the connection checks, you can make sure both sides of a connection will have the matching shape (barring edge cases like a block accepting both text and numbers, but I'll assume that can't apply to a block accepting both sprites and behaviors).

I hope this helps, sorry it took me a while to figure out the cause of this!

Reply:
Thanks so much! Yes - this makes sense. I was making things more complicated than they had to be!
I was thinking about the 'move [sprite] pixels [number] toward [location]' block which accepts a sprite, a number and a location. But at each specific connector, the type accepted will always be the same.